src/gsy_e/models/myco_matcher/myco_internal_matcher.py

```diff
@@ -32,16 +35,29 @@ class MycoInternalMatcher(MycoMatcherInterface):
     def __init__(self):
         super().__init__()
         self.match_algorithm = None
+        # TBD: Which matching algorithm is applied for which forward market type
+        self._forward_match_algorithms = {
+            AvailableMarketTypes.DAY_FORWARD: PayAsBidMatchingAlgorithm,
+            AvailableMarketTypes.WEEK_FORWARD: PayAsClearMatchingAlgorithm,
+            AvailableMarketTypes.MONTH_FORWARD: PayAsClearMatchingAlgorithm,
+            AvailableMarketTypes.YEAR_FORWARD: PayAsClearMatchingAlgorithm,
+        }
+        self._forward_market_counters = {
+            AvailableMarketTypes.DAY_FORWARD: DayForwardMarketCounter,
+            AvailableMarketTypes.WEEK_FORWARD: WeekForwardMarketCounter,
+            AvailableMarketTypes.MONTH_FORWARD: MonthForwardMarketCounter,
+            AvailableMarketTypes.YEAR_FORWARD: YearForwardMarketCounter
+        }
```

Comment:
If I get it right, the complication here is that the Myco matcher doesn't apply just one algorithm to all situations anymore (as it did before). Now it can instead use different algorithms depending on the type of exchange (spot/future/day-forward/month-forward/etc.). Is this correct?

As I mentioned in one of the previous PRs, I still think that this kind of change complicates too much the interface. Before, we only had `matcher.match_algorithm`, while now we've branched the logic of this class for two different scenarios ("normal" vs "forward" markets). This branching makes the class much more complex to understand and use in my opinion.
In fact, you can already see how the `match_recommendations()` method became way more complicated: we're passing a callable to it, and sometimes this callable is inside the class itself, while some other times it comes from a different class/algorithm.

I think that we should unify these two logics. Either we always use `self.match_algorithm` or we completely replace that and integrate it with the new logic you're implementing here. Ideally, there should be one way to deal with all types of "market types" and matching algorithms, without creating ad-hoc separate variables and logic for each of them.

I'm sorry if I can't be much more specific than this: as you see, there are several parts of the code involved and it's challenging to get a grasp of it without working on the code. But maybe there's another class emerging here and asking to be implemented.

P.S.: The current solution might also complicate the situations in which we're currently calling `matcher.match_algorithm.state`, because this doesn't work for future/forward matching algorithms.

Reply:
> If I get it right, the complication here is that the Myco matcher doesn't apply just one algorithm to all situations anymore (as it did before). Now it can instead use different algorithms depending on the type of exchange (spot/future/day-forward/month-forward/etc.). Is this correct?

Correct!

For the rest of your comment, I hear you and was also thinking about refactoring. But I first wanted to present a first implementation. IMO the "new class" approach makes sense. I will try to incorporate the `matcher.match_algorithm.state` issue, thangs for pointing me to it.

Reply:
I moved the forward market matching in its own class, that is initiated by the global singleton `bid_offer_matcher`. This will also only provided with the orders from the forward markets from the area, too.
I am not sure whether this is the most pretty solution, but it divides responsibilities better than before.
https://github.com/gridsingularity/gsy-e/pull/1463/commits/d079dc44ffc6167aa9b54c152b1d18726d00e6dc

We only have to deal with the `matcher.match_algorithm.state` once we want to plot the supply and demand curves for the future markets, so I suggest to deal with this later. For now there is no urge AFAIS.